Login and registration localisation

Feodor0090 opened the pull request:
Everything except the warning in `ScreenWarning`. I don't know what to do with it.

Feodor0090 committed:
Localise login form

Feodor0090 committed:
Localise registration window

Feodor0090 committed:
Reuse existing

Feodor0090 committed:
Localise caps lock warning

peppy commented:
Going forward, could you please limit yourself to one PR open at a time for localisation tasks? This will ensure we can keep up with review and give you feedback as we go.

peppy commented on osu.Game/Localisation/AccountCreationStrings.cs:
```diff
@@ -0,0 +1,72 @@
+        /// <summary>
+        /// ". Choose something long but also something you will remember, like a line from your favourite song."
+        /// </summary>
+        public static LocalisableString AfterCharactersLong =>
```
This won't work well for languages that aren't english. You'll need to use a formatting string with the "8 characters long" part as an argument.

Feodor0090 replied:
Are there existing examples how this should work?
I found this case which is quite similar, but it works with links, not custom drawables:

https://github.com/ppy/osu/blob/3f2077a527ade1380b77c9ac55ca2ec3ddd84616/osu.Game/Overlays/Settings/Sections/Input/TabletSettings.cs#L112-L117

I tried with splitting the string on place, but it looks ugly.

Joehuu replied:
I don't believe there are any. Also, the code above doesn't even work, `FormatText` and `GetLocalisedBindableString().Value` return a string, so the `LocalisableString` fallbacks to English.

Feodor0090 replied:
Am i right that we need some kind of `TextFlowContainer` which will track base localisable string with custom drawables inside it and recompose the text on language changes?

bdach replied:
I would not use `CustomizableTextContainer` here. At least not raw. If I'm thinking correctly what you were going to try and do (replace a placeholder with a `SpriteText`), it will not handle line breaking correctly.

Needs proper framework support somehow. I don't yet know how.

Feodor0090 committed:
Replace 2 strings with one formattable

Feodor0090 committed:
Merge branch 'master' into registration-loc

peppy committed:
Merge branch 'master' into registration-loc

peppy commented on osu.Game/Overlays/AccountCreation/ScreenEntry.cs:
```diff
@@ -132,14 +134,17 @@ private void load()
             textboxes = new[] { usernameTextBox, emailTextBox, passwordTextBox };
 
-            usernameDescription.AddText("This will be your public presence. No profanity, no impersonation. Avoid exposing your own personal details, too!");
+            usernameDescription.AddText(AccountCreationStrings.ThisWillBeYourPublic);
```
`UsernameDescription`

peppy commented on osu.Game/Overlays/AccountCreation/ScreenEntry.cs:
```diff
@@ -132,14 +134,17 @@ private void load()
 
-            emailAddressDescription.AddText("Will be used for notifications, account verification and in the case you forget your password. No spam, ever.");
-            emailAddressDescription.AddText(" Make sure to get it right!", cp => cp.Font = cp.Font.With(Typeface.Torus, weight: FontWeight.Bold));
+            emailAddressDescription.AddText(AccountCreationStrings.EmailUsage);
```
`EmailDescription1`

peppy commented on osu.Game/Overlays/AccountCreation/ScreenEntry.cs:
```diff
@@ -132,14 +134,17 @@ private void load()
-            emailAddressDescription.AddText("Will be used for notifications, account verification and in the case you forget your password. No spam, ever.");
-            emailAddressDescription.AddText(" Make sure to get it right!", cp => cp.Font = cp.Font.With(Typeface.Torus, weight: FontWeight.Bold));
+            emailAddressDescription.AddText(AccountCreationStrings.EmailUsage);
+            emailAddressDescription.AddText(AccountCreationStrings.MakeSureToGetIt, cp => cp.Font = cp.Font.With(Typeface.Torus, weight: FontWeight.Bold));
```
`EmailDescription2`

peppy commented on osu.Game/Overlays/AccountCreation/ScreenEntry.cs:
```diff
@@ -132,14 +134,17 @@ private void load()
-            passwordDescription.AddText("At least ");
-            characterCheckText = passwordDescription.AddText("8 characters long");
-            passwordDescription.AddText(". Choose something long but also something you will remember, like a line from your favourite song.");
+            string[] passwordReq = localisationManager.GetLocalisedBindableString(AccountCreationStrings.PasswordRequirements("{}")).Value.Split("{}");
```
What on earth is this.

peppy commented on osu.Game/Overlays/AccountCreation/ScreenWarning.cs:
```diff
@@ -101,13 +102,13 @@ private void load(OsuColour colours, TextureStore textures)
                         new SettingsButton
                         {
-                            Text = "Help, I can't access my account!",
+                            Text = AccountCreationStrings.HelpICantAccess,
```
`MultiAccountWarningHelp`

peppy commented on osu.Game/Overlays/AccountCreation/ScreenWarning.cs:
```diff
@@ -101,13 +102,13 @@ private void load(OsuColour colours, TextureStore textures)
                         new DangerousSettingsButton
                         {
-                            Text = "I understand. This account isn't for me.",
+                            Text = AccountCreationStrings.AccountIsntForMe,
```
`MultiAccountWarningAccept`

peppy reviewed (dismissed):
I'm not going to go through every case, but name the keys better. They should better explain the usage, not just use the default whatever.

peppy commented:
@Feodor0090 are you planning to apply the review for this?

Feodor0090 commented:
I want to bring o!f support for text chunks with differently formatted parts first. This is necessary for password requirement text, for example. I attempted to do it, but we decided that i chose a bad approach. Will return to this when I have free time.

bdach commented:
For visibility/posterity, the attempt mentioned above is https://github.com/ppy/osu-framework/pull/5632.

Joehuu committed:
Merge remote-tracking branch 'upstream/master' into pr/22239

Joehuu committed:
Revert blocking password requirement text localisation

Joehuu committed:
Apply key renaming suggestions

Joehuu committed:
Apply automated formatting changes

Joehuu committed:
Normalise source strings to sentence case

The casing of the "caps lock" tooltip wasn't changed in code, as tooltips should be ideally sentence-cased, see https://github.com/ppy/osu/pull/21765#issuecomment-1552378930.

Joehuu commented:
Have applied changes, as I believe this shouldn't be blocked by the password requirement text case and can be done later with all the other strings that need styling / multiple drawables.

Joehuu committed:
Remove unused load dependency

bdach commented on osu.Game/Overlays/AccountCreation/ScreenEntry.cs:
```diff
@@ -132,10 +133,10 @@ private void load()
 
-            emailAddressDescription.AddText("Will be used for notifications, account verification and in the case you forget your password. No spam, ever.");
-            emailAddressDescription.AddText(" Make sure to get it right!", cp => cp.Font = cp.Font.With(Typeface.Torus, weight: FontWeight.Bold));
+            emailAddressDescription.AddText(AccountCreationStrings.EmailDescription1);
+            emailAddressDescription.AddText(AccountCreationStrings.EmailDescription2, cp => cp.Font = cp.Font.With(Typeface.Torus, weight: FontWeight.Bold));
```
This I'm still unsure about. For now it's *probably* okay given we don't support any right-to-left scripts. But eventually this is gonna break. If you have to break a coherent string in half due to formatting then that's a bit of a giveaway that it may not localise correctly.

peppy replied:
Probably fine for now.
